packages/spatial/src/renderer/WebGLRendererSystem.test.tsx

```diff
@@ -0,0 +1,197 @@
+    setComponent(nestedVisibleEntity, EntityTreeComponent)
+    setComponent(invisibleEntity, SceneComponent, { children: [nestedVisibleEntity] })
+
+    //override addpass to test data without dependency on Browser
+    let addPassCount = 0
+    EffectComposer.prototype.addPass = () => {
+      addPassCount++
+    }
+  })
+
+  afterEach(() => {
+    return destroyEngine()
+  })
+
+  it('Test Background, Environment, Fog Components', async () => {
+    const { background, environment, fog, children } = getSceneParameters([rootEntity])
+    SystemDefinitions.get(WebGLRendererSystem)?.execute()
+    assert(background, 'background component exists')
+    const backgroundColor = background as Color
+    assert(
+      backgroundColor.r == 1 && backgroundColor.g == 1 && backgroundColor.b == 1,
+      'backgroud color was set correctly'
+    )
+
+    assert(environment, 'environment component exists')
+    assert(fog, 'fog component exists')
+  })
+
+  it('Test WebGL Reactors', async () => {
+    const webGLRendererSystem = SystemDefinitions.get(WebGLRendererSystem)
+    const RenderSystem = webGLRendererSystem?.reactor!
+    const tag = <RenderSystem />
+    const { rerender, unmount } = render(tag)
+
+    SystemDefinitions.get(WebGLRendererSystem)?.execute()
+
+    const engineRendererSettings = getMutableState(RendererState)
+    engineRendererSettings.renderMode.set(RenderModes.WIREFRAME)
+    engineRendererSettings.renderScale.set(2)
+    engineRendererSettings.qualityLevel.set(3)
+    engineRendererSettings.automatic.set(false)
+    engineRendererSettings.physicsDebug.set(true)
+    engineRendererSettings.avatarDebug.set(true)
+    engineRendererSettings.gridVisibility.set(true)
+    engineRendererSettings.nodeHelperVisibility.set(true)
+
+    await act(() => rerender(tag))
+
+    const camera = getComponent(rootEntity, CameraComponent)
+    const rendererComp = getComponent(rootEntity, RendererComponent)
+    const effectComposer = rendererComp.effectComposer
+    const passes = effectComposer?.passes.filter((p) => p.name === 'RenderPass') as any
+    const renderPass: RenderPass = passes ? passes[0] : undefined
+
+    const performanceState = getMutableState(PerformanceState)
+
+    assert(renderPass.overrideMaterial, 'change render mode')
+    assert(rendererComp.needsResize, 'change render scale')
+    assert(performanceState.cpuTier.value == 3, 'change quality level')
+    assert(camera.layers.isEnabled(ObjectLayers.PhysicsHelper), 'enable physicsDebug')
+    assert(camera.layers.isEnabled(ObjectLayers.AvatarHelper), 'enable avatarDebug')
+    assert(camera.layers.isEnabled(ObjectLayers.Gizmos), 'enable gridVisibility')
+    assert(camera.layers.isEnabled(ObjectLayers.NodeHelper), 'enable nodeHelperVisibility')
+
+    webGLRendererSystem?.execute()
+
+    assert(!rendererComp.needsResize, 'resize updated')
+    assert(
+      globalThis._scene.children.length == 1 && globalThis._scene.children[0].entity == visibleEntity,
```

Comment:
we should not rely on `globalThis._scene`, rather than assert the data put onto the _scene object, we should either change the implementation of the function that modifies it to be more functionally oriented (smaller composed functions that do less, and are more pure, such that they are more easily testable), or test the data some other way (such as overriding/mocking the render function and asserting the data passed to it)